File: src/main/java/net/alexwells/roomery/mechanic/roomholder/RoomHolderBlock.java
package net.alexwells.roomery.mechanic.roomholder;

import net.alexwells.roomery.Roomery;
import net.alexwells.roomery.gui.GuiEnum;
import net.alexwells.roomery.util.GuiUtils;
import net.alexwells.roomery.util.WorldUtils;
import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyBool;
import net.minecraft.block.properties.PropertyDirection;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.NonNullList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;

import javax.annotation.Nullable;

public class RoomHolderBlock extends Block {
    public static final String NAME = "room_holder";

    public static final PropertyDirection FACING = PropertyDirection.create("facing");
    // Changed from RoomHolderTileEntity
    public static final PropertyBool ACTIVE = PropertyBool.create("active");

    public RoomHolderBlock() {
        super(Material.IRON);

        setHardness(2);
        setRegistryName(Roomery.MOD_ID, NAME);
        setTranslationKey(getRegistryName().toString());
        setDefaultState(blockState.getBaseState()
                .withProperty(FACING, EnumFacing.NORTH)
                .withProperty(ACTIVE, false)
        );
    }

    @Override
    protected BlockStateContainer createBlockState() {
        return new BlockStateContainer(this, FACING, ACTIVE);
    }

    @Override
    public int getMetaFromState(IBlockState state) {
        return state.getValue(FACING).getIndex();
    }

    @Override
    public IBlockState getStateFromMeta(int meta) {
        return getDefaultState().withProperty(FACING, EnumFacing.byIndex(meta));
    }

    @Override
    public IBlockState getActualState(IBlockState state, IBlockAccess world, BlockPos pos)
    {
        RoomHolderTileEntity tile = WorldUtils.getTileEntity(world, pos);

        return state.withProperty(ACTIVE, tile != null && tile.isActive());
    }

    @Override
    public IBlockState getStateForPlacement(World world, BlockPos pos, EnumFacing facing, float hitX, float hitY, float hitZ, int meta, EntityLivingBase placer, EnumHand hand) {
        return getDefaultState().withProperty(FACING, EnumFacing.getDirectionFromEntityLiving(pos, placer));
    }

    @Override
    public boolean hasTileEntity(IBlockState p_hasTileEntity_1_) {
        return true;
    }

    @Nullable
    @Override
    public TileEntity createTileEntity(World world, IBlockState state) {
        return new RoomHolderTileEntity();
    }

    @Override
    public void getDrops(NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune) {
        super.getDrops(drops, world, pos, state, fortune);

        RoomHolderTileEntity tile = WorldUtils.getTileEntity(world, pos);

        if(tile == null || tile.getRoomCard() == null) {
            return;
        }

        drops.add(tile.getRoomCard());
    }

    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        // If sneaking, let the block be placed
        if(player.isSneaking()) {
            return false;
        }

        // If world is client sided, don't place a block
        if(world.isRemote) {
            return true;
        }

        RoomHolderTileEntity tile = WorldUtils.getTileEntity(world, pos);

        if(tile == null) {
            return false;
        }

        GuiUtils.openGui(player, GuiEnum.ROOM_HOLDER, world, pos);

        return true;
    }

    @Override
    public void onBlockPlacedBy(World worldIn, BlockPos pos, IBlockState state, EntityLivingBase placer, ItemStack stack) {
        super.onBlockPlacedBy(worldIn, pos, state, placer, stack);
    }
}

File: src/main/java/net/alexwells/roomery/mechanic/roomholder/RoomHolderBlock.kt
package net.alexwells.roomery.mechanic.roomholder

import net.alexwells.roomery.util.getTile
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.SoundType
import net.minecraft.block.material.Material
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.item.BlockItemUseContext
import net.minecraft.item.ItemStack
import net.minecraft.state.BooleanProperty
import net.minecraft.state.DirectionProperty
import net.minecraft.state.StateContainer
import net.minecraft.state.properties.BlockStateProperties
import net.minecraft.util.Direction
import net.minecraft.util.Hand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.BlockRayTraceResult
import net.minecraft.world.IBlockReader
import net.minecraft.world.World
import net.minecraft.world.storage.loot.LootContext
import net.minecraft.world.storage.loot.LootParameters
import net.minecraftforge.fml.network.NetworkHooks

object RoomHolderBlock : Block(
    Block.Properties
        .create(Material.IRON)
        .hardnessAndResistance(2f)
        .sound(SoundType.METAL)
) {
    // Sadly, we have to declare properties inside another object.
    // This is because if you declare them directly, Block()'s
    // constructor will call createBlockState(), which
    // will try to access these properties and get null.
    // These won't be initialized before Block()'s constructor
    // ends what it does.
    object Properties {
        val FACING: DirectionProperty = BlockStateProperties.HORIZONTAL_FACING
        val ACTIVE: BooleanProperty = BooleanProperty.create("active")
    }

    init {
        registryName = ROOM_HOLDER_RESOURCE
        defaultState = stateContainer.baseState
            .with(Properties.FACING, Direction.NORTH)
            .with(Properties.ACTIVE, false)
    }

    override fun fillStateContainer(builder: StateContainer.Builder<Block, BlockState>) {
        super.fillStateContainer(builder)

        builder.add(Properties.FACING, Properties.ACTIVE)
    }

    override fun hasTileEntity(state: BlockState?) = true

    override fun createTileEntity(state: BlockState?, world: IBlockReader?) = RoomHolderTileEntity()

    override fun getStateForPlacement(context: BlockItemUseContext): BlockState? {
        return defaultState.blockState.with(Properties.FACING, context.placementHorizontalFacing.opposite)
    }

    override fun getDrops(state: BlockState, context: LootContext.Builder): MutableList<ItemStack> {
        val drops = super.getDrops(state, context)

        val tile = context.get(LootParameters.BLOCK_ENTITY)
        if (tile is RoomHolderTileEntity && tile.roomCard != null) {
            drops.add(tile.roomCard)
        }

        return drops
    }

    override fun onBlockActivated(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockRayTraceResult
    ): Boolean {
        // If sneaking, let the block be placed
        if (player.isSneaking) {
            return false
        }

        // If world is client sided, don't place a block
        if (world.isRemote) {
            return true
        }

        val tile = world.getTile<RoomHolderTileEntity>(pos) ?: return false

        // todo: send chunk preview
        NetworkHooks.openGui(player as ServerPlayerEntity, tile, pos)

        return true
    }
}
